### PR TITLE
KmerCollection bug-fix, step-size, and tests.

### PR DESCRIPTION
**Pull request by Chrikomu**

### CODE REVIEW

File: src/kmer_collection.h [hunk 1]
```diff
@@ -136,47 +138,59 @@ class KmerCollection {
...
   }
 
-  KmerCollection::iterator end() {
-    return KmerIterator(this, m_kmerListSize);
-  }
-
- private:
-  void chooseGenerator() {
+  inline void chooseGenerator() {
     if(m_compress == false) switch(m_kmerSize) {
```
Comment (AndreasDahl) [inline src/kmer_collection.h]: Is `!m_compress` not preferred?

File: src/kmer_collection.h [hunk 2]
```diff
@@ -107,11 +107,11 @@ class KmerCollection {
...
-    m_kmerListSize(seqEntry->seq().size() - kmerSize + 1),
+    m_stepSize(stepSize),
+    m_scoreMin(scoreMin),
+    m_kmerListSize(0),
     m_kmerList(nullptr),
-    m_usedGenerator(&KmerCollection::generateKmersN),
+    m_usedGenerator(&KmerCollection::generateKmerN),
     m_compress(compressHint)
   {
     if(m_compress == false && kmerSize > 8) {
```
Comment (AndreasDahl) [inline src/kmer_collection.h]: Left side can be replaced with `!m_compress`

File: src/kmer_collection.h (hunk 1 above)
Comment (maasha) [inline src/kmer_collection.h]: I vote !m_compress

File: src/kmer_collection.h (hunk 2 above)
Comment (maasha) [inline src/kmer_collection.h]: +1

File: test/test_kmer_collection.cc [hunk 3]
```diff
@@ -43,10 +43,60 @@ TEST_CASE("Basic functionalities of kmer_collection", "[kmer_collection]") {
...
     REQUIRE(itEnd == itEnd);
   }
+
+  SECTION("Size returns correctly") {
+    REQUIRE(collection.size() == 20);
+  }
+}
+
+
+TEST_CASE("Unique kmers", "[kmer_collection]") {
```
Comment (AndreasDahl) [inline test/test_kmer_collection.cc]: It is not immediately obvious what this test case tests. Please add some comments or split it up into logical parts

File: test/test_kmer_collection.cc [hunk 4]
```diff
@@ -43,10 +43,60 @@ TEST_CASE("Basic functionalities of kmer_collection", "[kmer_collection]") {
     REQUIRE(itBegin != itEnd);
   }
 
-  SECTION("comparison for equality") {
+  SECTION("Comparison for equality 1") {
     REQUIRE(itBegin == itBegin);
     REQUIRE(itEnd == itEnd);
```
Comment (AndreasDahl) [inline test/test_kmer_collection.cc:50]: Are these `REQUIRE` not [tautological](https://en.wikipedia.org/wiki/Tautology_%28logic%29)?

File: src/kmer_collection.h [hunk 5]
```diff
@@ -136,47 +165,70 @@ class KmerCollection {
...
+      default: m_usedGenerator = &KmerCollection::generateKmerN; break;
     }
   }
 
   // This assumes the next 8 chars is readable
-  uint_fast64_t generateKmers8(const char* ptr) {
-    uint_fast64_t kmer = *(const uint_fast64_t*)ptr;
+  uint_fast64_t generateKmer8(const char* ptr) const {
+    const uint_fast64_t mask = 0x0606060606060606;
+    uint_fast64_t kmer = *(const uint_fast64_t*)ptr & mask;
```
Comment (AndreasDahl) [inline src/kmer_collection.h]: Don't use c-style casts and don't assume `uint_fast64_t` is guaranteed to be 64 bit.

File: test/test_kmer_collection.cc (hunk 4 above)
Comment (AndreasDahl) [inline test/test_kmer_collection.cc:50]: @ahovgaard i see you introduced these tests. Do they hold any value?

Comment (ahovgaard) [inline test/test_kmer_collection.cc:50]: Did I? I thought I just included them from an earlier commit/version.
I suppose they don't.

Comment (AndreasDahl) [inline test/test_kmer_collection.cc:50]: They may test the equality operator...

Comment (maasha) [inline test/test_kmer_collection.cc:50]: So test cases and sections should always describe:
1. What is being tested
2. The argument(s)
3. The expected outcome of the test

File: src/kmer_collection.h [hunk 6]
```diff
@@ -126,57 +153,76 @@ class KmerCollection {
...
+      default: m_usedGenerator = &KmerCollection::generateKmerN; break;
     }
   }
 
   // This assumes the next 8 chars is readable
-  uint_fast64_t generateKmers8(const char* ptr) {
-    uint_fast64_t kmer = *(const uint_fast64_t*)ptr;
+  uint_fast64_t generateKmer8(const char* ptr) const {
+    const uint_fast64_t mask = 0x0606060606060606;
+    uint64_t kmer = *(const uint64_t*)ptr & mask;
```
Comment (AndreasDahl) [inline src/kmer_collection.h:218]: This part is still extremely hacky, but i do not see a way to make it less hacky and still keep the speed.

**Comment by AndreasDahl** [comment]
I have done all the improvements i can think of with @Chrikomu now. A third party should review and possibly merge this pull request.

**Comment by AndreasDahl** [comment]
@ahovgaard @RasmusFonseca @maasha 
Someone please review this.

**Comment by ahovgaard** [comment]
I will look through it now.

File: src/kmer_collection.h [hunk 7]
```diff
@@ -126,57 +153,76 @@ class KmerCollection {
...
   KmerCollection::iterator begin() {
+    generateKmers();
+    return KmerIterator(this, 0);
+  }
+
+  /**
+  * \return Iterator to the element after KmerCollection's last element.
+  */
+  inline KmerCollection::iterator end() {
+    return KmerIterator(this, m_kmerListSize);
```
Comment (ahovgaard) [inline src/kmer_collection.h:185]: What if `m_kmerListSize` is still 0 when calling `end()`?
Shouldn't we also `generateKmers()` here?

**Pull request merged**